controllers/machinehealthcheck_controller.go

```diff
@@ -2,61 +2,504 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/pointer"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	"github.com/openshift/api/machine/v1beta1"
 
 	"github.com/medik8s/node-healthcheck-operator/controllers/cluster"
+	"github.com/medik8s/node-healthcheck-operator/controllers/featuregates"
 	"github.com/medik8s/node-healthcheck-operator/controllers/mhc"
+	"github.com/medik8s/node-healthcheck-operator/controllers/resources"
+	"github.com/medik8s/node-healthcheck-operator/controllers/utils"
+	"github.com/medik8s/node-healthcheck-operator/metrics"
+)
+
+const (
+	defaultNodeStartupTimeout = 10 * time.Minute
+)
+
+var (
+	// We allow users to disable the nodeStartupTimeout by setting the duration to 0.
+	disabledNodeStartupTimeout = metav1.Duration{Duration: 0}
 )
 
 // MachineHealthCheckReconciler reconciles a MachineHealthCheck object
 type MachineHealthCheckReconciler struct {
 	client.Client
-	Log                         logr.Logger
-	Scheme                      *runtime.Scheme
-	Recorder                    record.EventRecorder
-	ClusterUpgradeStatusChecker cluster.UpgradeChecker
-	MHCChecker                  mhc.Checker
+	Log                            logr.Logger
+	Recorder                       record.EventRecorder
+	ClusterUpgradeStatusChecker    cluster.UpgradeChecker
+	MHCChecker                     mhc.Checker
+	FeatureGateMHCControllerEvents <-chan event.GenericEvent
```

Comment:
I think we can get rid of this variable and instead have initialized inside the feature gate accessor

Reply:
you mean like this?
https://github.com/medik8s/node-healthcheck-operator/pull/224/commits/db541567c781ab9a194a01a2c15e057abd33c204